### PR TITLE
Merge next into master, include both Gatsby.js and Next.js

### PR DESCRIPTION
**Pull request by humphd**
- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [x] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description

I want to make it easier to run the Gatsby and Next frontends side-by-side.  The current branch-based separation makes it really hard to switch back and forth with dependencies.  To do that, I've merged the `next` branch into `master`, and created a temporary parallel front-end structure:

```
└──frontend
      └──next        
      └──gatsby
```

Our default front-end continues to be Gatsby, and all references to `src/frontend` have been updated to `src/frontend/gatsby`.

I've adjusted our scripts to work for both next and gatsby, and added temporary docs in `src/frontend/README.md`.

I've renamed the env variable `PROXY_GATSBY` To `PROXY_FRONTEND`, since it will work for both cases.

I've cleaned up the next.js deps and eslint config to be more accurate and up-to-date.

I've separated the front-end specific bits of `.gitignore` into each of the frontend dirs.

I think everything is working, but check my math please.

### CODE REVIEW

File: src/frontend/next/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    },
+  },
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    ecmaVersion: 2020,
+    sourceType: 'module',
+  },
+  extends: [
+    'eslint:recommended',
+    'airbnb-base',
```
Comment (manekenpix) [inline src/frontend/next/.eslintrc.js:20]: Isn't `airbnb-base` the one that comes without React plugins?
From `airbnb-base` npm:
>This package provides Airbnb's base JS .eslintrc (without React plugins) as an extensible shared config.

Also, from the `airbnb` npm page:
>If you don't need React, see eslint-config-airbnb-base.

Comment (humphd) [reply src/frontend/next/.eslintrc.js:20]: Thanks, I'll fix.

**Comment by humphd** [comment]
This is ready for review.  I've streamlined the eslint stuff so we only have a single code base, but it works across node, JS, and TS.  I had to fix a bunch of existing lint issues that were hiding in our code.

**Review by c3ho — DISMISSED** [review]
Tested with `npm run develop` was able to run it with local backend. All expected front-end functionality was still working. Also ran `npm run develop:next` and it built successfully

**Comment by humphd** [comment]
For testing this, I'll up-level the docs that are buried in this PR:

---------------------

# Telescope Frontend Development

## Introduction

Telescope is in the process of porting the front-end from [GatsbyJS](https://www.gatsbyjs.org/)
to [Next.js](https://nextjs.org), see https://github.com/Seneca-CDOT/telescope/issues/1316.

To facilitate this conversion, we are hosting both front-ends in the `src/frontend` folder. For
the time being, the root npm scripts will default to run the GatsbyJS front-end. However, when
we finish the conversion, we'll switch that to Next.js and delete the GatsbyJS code.

NOTE: if you are only working on the front-end code, set `PROXY_FRONTEND=1` in your `.env` to use
the development server (https://dev.telescope.cdot.systems) as your back-end.

## Running GatsbyJS

From the root of the project, you can run a number of GatsbyJS specific npm scripts:

1. [`npm run develop`](https://www.gatsbyjs.org/docs/gatsby-cli/#develop) to start the development server
1. [`npm run build`](https://www.gatsbyjs.org/docs/gatsby-cli/#build) to build the site in `src/frontend/gatsby/public` for production
1. [`npm run serve`](https://www.gatsbyjs.org/docs/gatsby-cli/#serve) to serve the production site in `src/frontend/gatsby/public` for testing
1. [`npm run clean`](https://www.gatsbyjs.org/docs/gatsby-cli/#clean) to delete generated build files and folders

## Running Next.js

From the root of the project, you can run a number of GatsbyJS specific npm scripts:

1. [`npm run develop:next`](https://nextjs.org/docs/api-reference/cli#development) to start the development server
1. [`npm run build:next`]() to build the site in `src/frontend/next/public` for production
1. [`npm run serve:next`](https://nextjs.org/docs/api-reference/cli#production) to serve the production site in `src/frontend/next/.next` for testing
```

**Review by manekenpix — CHANGES_REQUESTED** [review]
In `gatsby-config.js`, line 7, it needs an extra `../`:

```js
path: path.join(__dirname, '../../../', '.env'),
```

**Comment by humphd** [comment]
@manekenpix great catch, thanks for reviewing this so closely.  I'll correct and rebase soon to deal with these new changes.

**Comment by humphd** [comment]
Pushed a fix for the gatsby-config.js issue.  I think I'll wait for approval to do a rebase at the end, since it's going to be a bit of work.

**Comment by chrispinkney** [comment]
> NOTE: if you are only working on the front-end code, set `PROXY_FRONTEND=1` in your `.env` to use
> the development server (https://dev.telescope.cdot.systems) as your back-end.

Also, `API_URL` should be set to `API_URL=https://dev.telescope.cdot.systems` as I was getting no timeline until this was set. Might be an obvious thing that I just missed though, haha.

**Comment by chrispinkney** [comment]
Tested all the above commands using your remote, runs both Gatsby and Next start page; LGTM. (_always wanted to say that_)

**Review by cindyledev — APPROVED** [review]
I've tested this using Docker on Windows and I'm able to run both Gatsby and Next frontends :D

**Comment by humphd** [comment]
This is pretty close to being ready to merge.  When I do, it's going to affect all of these open PRs:

- #1368 @manekenpix 
- #1372 @lixiaoqity 
- #1370 @phast184 
- #1376 @NathanPang001 
- #1414 @abhaseen 
- #1433 @PedroFonsecaDEV 
- #1440 @jiyoungsin 
- #1445 @c3ho 
- #1452 @tianlangwu 

Those of you working on anything in `src/frontend/*` will have to rebase after this goes in, and your code will need to move to `src/frontend/gatsby/*`.

I'm giving you all a heads up today, in case you want to get your stuff merged soon, so the rebase is my problem.

@c3ho, @manekenpix can I get another r? on this.

**Comment by manekenpix** [comment]
@humphd Vercel failed, and there are still conflicts. Are we ok merging with the conflicts?

**Comment by humphd** [comment]
Vercel is just wanting me to approve the changes to the config (happens every time I push), and I'll fix merge conflicts once it's approved.

**Review by manekenpix — APPROVED** [review]

**Review by c3ho — APPROVED** [review]

**Comment by humphd** [comment]
I've now merged this on `master` (see https://github.com/Seneca-CDOT/telescope/commit/6195af3fb3b9a277ca52e24a72c95e765be57791) and it's quite possible I've broken things.  We'll see how it does on https://dev.telescope.cdot.systems in a few mins, and adjust as needed.

I the mean time, I wanted to mention a few things to everyone that has open PRs touching the frontend (@manekenpix, @lixiaoqity, @phast184, @NathanPang001, @abhaseen, @PedroFonsecaDEV, @jiyoungsin, @c3ho, @tianlangwu):

1. you should update your `master` branch `git checkout master` and `git pull upstream master` to get this change.
1. you should run `npm install` again
1. you should rebase your branch on master `git checkout your-branch` and `git rebase master`.  Fix any conflicts.
1. when you finish the rebase, confirm that you don't have new code in `src/frontend/*` and that all your changes are in `src/frontend/gatsby/*` (i.e., all frontend code has been moved to `gatsby/`).

If you have problems or need help, let us know in your PR.

**Pull request merged**